File: shared-types/src/types/practice.ts
import { BaseEntity, PracticeType, PracticeSessionStatus, PracticeGenerateStatus, AnswerStatus } from './common';
import { Student } from './user';
import { Textbook, Unit, Question } from './education';

// ===== 练习会话相关 =====

/**
 * 练习会话
 */
export interface PracticeSession extends BaseEntity {
  id: number;
  student_id: string;
  session_type: PracticeType;
  target_id?: number; // 单元ID或日期（如 20241123）
  textbook_id?: number;
  question_count: number;
  answer_count: number;
  correct_count: number;
  status: PracticeSessionStatus;
  generate_status: PracticeGenerateStatus;
  start_time: number;
  end_time?: number;

  // 关联数据
  student?: Student;
  textbook?: Textbook;
  unit?: Unit;
  questions?: Question[];
  answers?: PracticeAnswer[];
  report?: PracticeReport;
}

/**
 * 创建练习请求
 */
export interface CreatePracticeRequest {
  type: PracticeType;
  textbook_id: number;
  unit_id?: number;
  target_date?: string; // 格式: YYYYMMDD
  question_count?: number;
  difficulty?: string;
  knowledge_ids?: number[]; // 指定知识点
}

/**
 * 练习答案记录
 */
export interface PracticeAnswer extends BaseEntity {
  id: number;
  session_id: number;
  question_id: number;
  question_order: number; // 题目在练习中的顺序
  text_answer?: string; // 文本答案
  status: AnswerStatus; // 答题状态
  time_spent: number; // 耗时（秒）
  submit_time?: number; // 提交时间
  audio_answer?: string; // 音频答案（OSS 存储路径）

  // 关联数据
  question?: Question;
}

/**
 * 提交答案请求
 */
export interface SubmitAnswerRequest {
  session_id: number;
  question_id: number;
  answer: string;
  time_spent: number; // 答题耗时（秒）
  is_audio_answer?: boolean; // 是否为音频回答
  audio_data?: string; // 音频 OSS 存储路径
  audio_match?: boolean; // 音频理解结果：是否匹配题目要求
  audio_reason?: string; // 音频理解结果：原因说明
  audio_suggestion?: string; // 音频理解结果：改进建议
}

/**
 * 提交答案响应
 */
export interface SubmitAnswerResponse {
  is_correct: boolean;
  correct_answer: string;
  user_answer?: string; // 用户答案（可能是ASR识别后的文本）
  analysis?: string; // 错题分析
  explanation?: string; // 题目解析
  session_progress: {
    answer_count: number;
    correct_count: number;
    total_count: number;
    status: PracticeSessionStatus;
  };
}

// ===== 错题记录 =====

/**
 * 错题记录
 */
export interface WrongRecord extends BaseEntity {
  id: number;
  student_id: string;
  question_id: number;
  session_id: number;
  unit_id?: number;
  knowledge_ids?: number[]; // 错题涉及的知识点
  knowledge_names?: string[];
  textbook_id?: number;
  user_answer?: string;
  correct_answer?: string;
  analysis?: string; // 错题分析
  time_spent?: number;
  is_corrected: boolean; // 是否已订正
  corrected_time?: number;
  review_count: number; // 复习次数
  mastery_level: number; // 掌握程度 1-5

  // 关联数据
  question?: Question;
  session?: PracticeSession;
}

// ===== 练习报告 =====

/**
 * 练习报告
 */
export interface PracticeReport extends BaseEntity {
  id: number;
  session_id: number;
  student_id: string;
  total_questions: number;
  correct_questions: number;
  total_time: number; // 总耗时（秒）
  overall_score: number; // 总得分
  accuracy_rate: number; // 正确率

  // 能力评估（主要用于assessment）
  current_ability?: number; // 当前能力值（-3到+3）
  ability_level?: string; // 能力等级
  percentile?: number; // 百分位排名
  confidence?: number; // 置信度

  // 详细分析（JSON格式）
  knowledge_scores?: Record<string, number>; // 知识点掌握情况
  question_distribution?: Record<string, number>; // 题目来源分布
  ability_breakdown?: Record<string, any>; // 能力分解
  learning_speed?: number; // 学习速度
  consistency?: number; // 稳定性
  strengths?: string[]; // 优势知识点
  weaknesses?: string[]; // 薄弱知识点
  recommendations?: string[]; // 学习建议

  // 关联数据
  session?: PracticeSession;
}

/**
 * 能力评估详情
 */
export interface AbilityAssessment {
  overall_score: number;
  ability_level: string;
  percentile: number;
  confidence: number;
  subject_scores: Record<string, number>;
  knowledge_mastery: Record<string, {
    score: number;
    level: string;
    improvement_needed: boolean;
  }>;
  learning_trends: {
    speed: number;
    consistency: number;
    improvement_rate: number;
  };
  next_step_recommendations: string[];
}

// ===== 学习统计 =====

/**
 * 学生学习统计
 */
export interface LearningStatistics {
  student_id: string;
  period: string; // 统计周期：daily/weekly/monthly

  // 基础统计
  total_practices: number;
  total_questions: number;
  total_time: number; // 总学习时间（秒）
  average_accuracy: number;
  average_time_per_question: number; // 平均每题时间（秒）

  // 按类型统计
  practice_type_stats: Record<PracticeType, {
    count: number;
    questions: number;
    accuracy: number;
    time: number;
  }>;

  // 知识点掌握度
  knowledge_mastery: Record<string, {
    total_attempts: number;
    correct_attempts: number;
    accuracy: number;
    mastery_level: number;
  }>;

  // 学习趋势
  daily_progress: Array<{
    date: string;
    questions: number;
    accuracy: number;
    time: number;
  }>;

  // 薄弱点分析
  weak_areas: Array<{
    knowledge_name: string;
    accuracy: number;
    recommended_actions: string[];
  }>;
}

File: apps/server/shared/core/database/practice.py
"""
练习相关模型

包含 Practice、PracticeAnswer、PracticeReport

练习类型分类：
- ability_practice: 能力练习 - 基于原子能力 code 生成题目
- unit_practice: 单元练习 - 基于单元 ID 生成题目
"""

import uuid
from typing import TYPE_CHECKING, Optional

from .base import (
    BaseModel,
    Integer,
    Mapped,
    String,
    Text,
    mapped_column,
    now,
    relationship,
)

if TYPE_CHECKING:
    from .question import Question
    from .student import Student


def generate_session_id() -> str:
    """生成练习会话 ID (UUID v4)"""
    return str(uuid.uuid4())


class Practice(BaseModel):
    """练习表

    练习类型 (practice_type):
    - ability_practice: 能力练习 - 基于原子能力 code 列表生成
    - unit_practice: 单元练习 - 基于单元 ID 生成

    生成状态 (generate_status):
    - 0: 生成中
    - 1: 已完成
    - -1: 生成失败
    """

    __tablename__ = "ah_practice"

    # 使用 UUID v4 作为主键
    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=generate_session_id, comment="会话ID (UUID v4)"
    )
    student_id: Mapped[str] = mapped_column(String(255), index=True, comment="学生ID")
    practice_type: Mapped[str] = mapped_column(
        String(50), index=True, comment="练习类型: ability_practice/unit_practice"
    )

    # 练习参数字段
    subject: Mapped[Optional[str]] = mapped_column(
        String(50), nullable=True, index=True, comment="科目"
    )
    grade: Mapped[Optional[int]] = mapped_column(Integer, nullable=True, index=True, comment="年级")
    ability_code: Mapped[Optional[str]] = mapped_column(
        String(255), nullable=True, index=True, comment="原子能力代码"
    )
    unit_id: Mapped[Optional[int]] = mapped_column(
        Integer, nullable=True, index=True, comment="单元ID"
    )

    # 题目统计
    question_count: Mapped[int] = mapped_column(default=0, comment="题目数量")
    answer_count: Mapped[int] = mapped_column(default=0, comment="回答数量")
    correct_count: Mapped[int] = mapped_column(default=0, comment="正确数量")

    # 状态字段
    status: Mapped[int] = mapped_column(
        default=0, index=True, comment="未开始: 0, 进行中: 1, 已完成: 2, 已废弃: 3"
    )
    generate_status: Mapped[int] = mapped_column(
        default=0, index=True, comment="生成中: 0, 已完成: 1, 生成失败: -1"
    )
    generate_time: Mapped[Optional[int]] = mapped_column(nullable=True, comment="生成耗时(秒)")
    start_time: Mapped[int] = mapped_column(default=now, comment="开始时间")
    end_time: Mapped[Optional[int]] = mapped_column(nullable=True, comment="结束时间")

    # 时间戳
    create_time: Mapped[int] = mapped_column(default=now, comment="创建时间")
    update_time: Mapped[int] = mapped_column(default=now, onupdate=now, comment="更新时间")

    # 关联关系
    student: Mapped["Student"] = relationship(
        "Student",
        primaryjoin="foreign(Practice.student_id) == Student.id",
        lazy="select",
    )


class PracticeAnswer(BaseModel):
    """答题记录表"""

    __tablename__ = "ah_practice_answer"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)

    # 基础关联字段 - session_id 改为 str 类型 (UUID v4)
    session_id: Mapped[str] = mapped_column(String(36), index=True, comment="会话ID (UUID v4)")
    question_id: Mapped[str] = mapped_column(String(255), index=True, comment="题目ID")
    student_id: Mapped[str] = mapped_column(String(255), index=True, comment="学生ID")
    question_order: Mapped[int] = mapped_column(comment="题目顺序")

    # 答题信息
    text_answer: Mapped[str] = mapped_column(Text, nullable=True, comment="学生答案")
    status: Mapped[int] = mapped_column(default=0, comment="答题状态: 0-未答 1-正确 2-错误")
    time_spent: Mapped[int] = mapped_column(default=0, comment="耗时(秒)")
    submit_time: Mapped[int] = mapped_column(nullable=True, comment="提交时间")

    # 错题相关字段
    correct_answer: Mapped[str] = mapped_column(Text, nullable=True, comment="正确答案")
    analysis: Mapped[str] = mapped_column(Text, nullable=True, comment="错题分析")
    is_corrected: Mapped[int] = mapped_column(default=0, comment="是否已订正 0-未订正 1-已订正")
    corrected_time: Mapped[int] = mapped_column(nullable=True, comment="订正时间")

    # 时间字段
    create_time: Mapped[int] = mapped_column(default=now, comment="创建时间")
    update_time: Mapped[int] = mapped_column(default=now, onupdate=now, comment="更新时间")

    question: Mapped["Question"] = relationship(
        "Question",
        primaryjoin="foreign(PracticeAnswer.question_id) == Question.id",
        lazy="joined",
    )


class PracticeReport(BaseModel):
    """练习报告表"""

    __tablename__ = "ah_practice_report"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    session_id: Mapped[str] = mapped_column(
        String(36), unique=True, index=True, comment="会话ID (UUID v4)"
    )
    student_id: Mapped[str] = mapped_column(String(255), comment="学生ID")

    # 总体统计
    total_questions: Mapped[int] = mapped_column(default=0, comment="题目数量")
    correct_questions: Mapped[int] = mapped_column(default=0, comment="正确数量")
    total_time: Mapped[int] = mapped_column(default=0, comment="总耗时(秒)")
    overall_score: Mapped[float] = mapped_column(default=0.0, comment="总得分")

    # 综合评估
    current_ability: Mapped[float] = mapped_column(default=0.0, comment="当前能力值（-3到+3）")
    confidence: Mapped[float] = mapped_column(default=0.0, comment="置信度")
    ability_level: Mapped[str] = mapped_column(String(50), default="", comment="能力等级")
    percentile: Mapped[int] = mapped_column(default=0, comment="百分位排名")

    # 详细分析 - JSON格式
    knowledge_scores: Mapped[str] = mapped_column(Text, default="{}", comment="知识点掌握情况")
    question_distribution: Mapped[str] = mapped_column(Text, default="{}", comment="题目来源分布")
    ability_breakdown: Mapped[str] = mapped_column(Text, default="{}", comment="能力分解（按难度）")
    learning_speed: Mapped[float] = mapped_column(default=0.0, comment="学习速度")
    consistency: Mapped[float] = mapped_column(default=0.0, comment="稳定性")

    # 建议 - JSON格式
    strengths: Mapped[str] = mapped_column(Text, default="[]", comment="优势")
    weaknesses: Mapped[str] = mapped_column(Text, default="[]", comment="薄弱点")
    recommendations: Mapped[str] = mapped_column(Text, default="[]", comment="学习建议")

    create_time: Mapped[int] = mapped_column(default=now, comment="创建时间")


__all__ = [
    "Practice",
    "PracticeAnswer",
    "PracticeReport",
    "generate_session_id",
]
